scaper/core.py

```diff
@@ -1211,25 +1284,34 @@ def _instantiate_event(self, event, isbackground=False,
                             event_duration),
                         ScaperWarning)
 
-        # determine source time
-        source_time = -np.Inf
-        while source_time < 0:
-            source_time = _get_value_from_dist(event.source_time)
-
-        # Make sure source time + event duration is not greater than the
-        # source duration, if it is, adjust the source time (i.e. duration
-        # takes precedences over start time).
-        if source_time + event_duration > source_duration:
-            old_source_time = source_time
-            source_time = source_duration - event_duration
-            if not disable_instantiation_warnings:
+        # Modify event.source_time so that sampling from the source time distribution
+        # stays within the bounds of the audio file - event_duration. This allows users 
+        # to sample from anywhere in a source file without knowing the exact duration 
+        # of every source file. Only modify if label is not in protected labels.
+        if label not in self.protected_labels:
+            modified_source_time, warn = _ensure_valid_source_time_tuple_bounds(
+                event.source_time, source_duration,  event_duration
+            )
+            
+            # determine source time and also check again just in case (for normal dist).
+            # if it happens again, just use the old method.
+            source_time = -np.Inf
+            while source_time < 0:
+                source_time = _get_value_from_dist(modified_source_time)
+                if source_time + event_duration > source_duration:
```

Comment:
Why is this check still necessary? Assuming the `_ensure...` function makes sure the `source_time` tuple is satisfiable, this test is redundant, right? I'm not keen on having `source_time` modified in multiple different places to fix the same issue, can make the program logic hard to keep track of. Unless there's a good reason to keep this test I think we should remove it.

Reply:
The check is still needed because normal distributions can still result in negative times.

Reply:
That's caught by `while source_time < 0:`. But I don't see why we need the `if source_time + event_duration > source_duration:` check.

Reply:
Oh, right you are. I didn't read one line down haha.